program/src/processor/withdraw_inactive_stake.rs

```diff
@@ -17,10 +26,150 @@ use crate::instruction::accounts::{Context, WithdrawInactiveStakeAccounts};
+        ProgramError::UninitializedAccount,
+        "config",
+    );
+
+    // stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - derivation must match (validates the config account)
+
+    require!(
+        ctx.accounts.stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "stake"
+    );
+
+    let mut stake_data = ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(&mut stake_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    let (derivation, _) =
+        find_stake_pda(&stake.validator_vote, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // stake authority
+    // - must be a signer
+    // - must match the authority on the stake account
+
+    require!(
+        ctx.accounts.stake_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake authority",
+    );
+
+    require!(
+        ctx.accounts.stake_authority.key == &stake.authority,
+        StakeError::InvalidAuthority,
+        "stake authority",
+    );
+
+    // vault authority
+    // - derivation must match
+
+    let bump = [config.vault_authority_bump];
+    let vault_signer = create_vault_pda(ctx.accounts.config.key, &bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &vault_signer,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &bump);
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    let mint_data = ctx.accounts.mint.try_borrow_data()?;
+    let mint = PodStateWithExtensions::<PodMint>::unpack(&mint_data)?;
+    let decimals = mint.base.decimals;
+
+    // Update the config and stake account data.
+
+    require!(
+        amount <= stake.inactive_amount,
+        StakeError::NotEnoughInactivatedTokens,
+        "amount"
+    );
+
+    stake.inactive_amount = stake
+        .inactive_amount
+        .checked_sub(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    config.token_amount_delegated = config
+        .token_amount_delegated
+        .checked_sub(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
```

Comment:
Maybe this is clarified later, but `amount` should be removed from `token_amount_delegated` during `InactivateStake`, and not during withdrawal

Reply:
Makes sense, but this has an implication on the `Slash` instruction ability to slash inactive stake. I think we have a couple of options:
1. In the current design, we can slash inactive stake since it is still in the system (the `token_amount_delegated` "knows" about them). Spoke to Joe earlier, and one way to keep this capability is to create a `token_amount_inactive` on the `Config` account to track inactive stake which has not been withdrawn.
2. We don't slash inactive stake, so we modify the `Slash` instruction to only go through `active` and `deactivating` stake.

Reply:
I would go with number 2.

Imagine if you have staked SOL:

* you deactivate it
* you wait, it becomes inactive, but you don't withdraw
* the validator you were staked to commits a slashable offense
* you get slashed

That would be *very* bad

Reply:
Is there a risk that someone keeps track of the governance vote and as soon as it sees that the validator could be slashed, inactivate its stake? Not sure if there are guarantees that the voting period will always be shorter than the cooldown period.

I am thinking the scenario where:
* validator commits a slashable offence
* slash proposal begin voting phase
* every staker inactive their stake
* slash proposal is executed
* there are no tokens left to slash

Slashing inactive tokens does not prevent the scenario above, but adds an extra step.

Reply:
There's a 30-day cooldown period for deactivation, and it's only possible to deactivate 10% at a time, so even the voting period takes 30 days (which shouldn't happen), they'll still lose 90% of their stake

Reply:
Cool, sounds good. I will update the logic.